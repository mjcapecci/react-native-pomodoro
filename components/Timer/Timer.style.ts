import { StyleSheet } from 'react-native'

export default StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: '#000',
  },
  timeHeaderContainer: {
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    height: 100,
  },
  timeHeader: {
    fontSize: 84,
    color: '#fff',
    marginBottom: 5,
    fontVariant: ['tabular-nums'],
    borderLeftWidth: 4,
    borderRightWidth: 4,
  },
  starContainer: {
    display: 'flex',
    flexDirection: 'row',
  },
  star: {
    margin: 3,
  },
  actionButton: {
    marginTop: 60,
    background: 'none',
  },
  skipButton: {
    marginTop: 30,
    background: 'none',
  },
  hideSkip: {
    display: 'none',
  },
  loadingCaption: {
    color: '#fff',
    marginBottom: 25,
    fontSize: 24,
  },
})
