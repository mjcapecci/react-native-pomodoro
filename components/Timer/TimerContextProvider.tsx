import React, { createContext, useCallback, useEffect, useState } from 'react'
import { getNextRound, getRoundType, getSecondsReset } from './helpers/timerHelpers'
import { scheduleNotification, cancelAllNotifications } from '../Notifications/notificationManager'
import AsyncStorage from '@react-native-async-storage/async-storage'
import { AppState, AppStateStatus } from 'react-native'
import getTimeRemaining, { getRoundData } from './helpers/getTimeRemaining'
import { RoundData, RoundType } from '../../types'
import addRecord from '../../data_layer/addRecord'
import shouldAddRecord from './helpers/shouldAddRecord'
import { useMutation, useQueryClient } from '@tanstack/react-query'

export interface TimerContextProps {
  enabled: boolean
  startButtonEnabled: boolean
  timerActive: boolean
  roundNumber: number
  roundType: string
  secondsLeft: number
  appStateVisible: boolean
  advanceRound: () => void
  startRound: () => void
  stopRound: () => void
  showConfirmationModal: boolean
  setShowConfirmationModal: (show: boolean) => void
}

const TimerContext = createContext<TimerContextProps>({
  enabled: true,
  startButtonEnabled: true,
  timerActive: false,
  roundNumber: 0,
  roundType: 'work',
  secondsLeft: 1500,
  appStateVisible: true,
  advanceRound: () => null,
  startRound: () => null,
  stopRound: () => null,
  showConfirmationModal: false,
  setShowConfirmationModal: () => null,
})

interface TimerContextProviderProps {
  children: JSX.Element
}

function TimerContextProvider({ children }: TimerContextProviderProps): JSX.Element {
  const [enabled, setEnabled] = useState(true)
  const [startButtonEnabled, setStartButtonEnabled] = useState(true)
  const [timerActive, setTimerActive] = useState(false)
  const [roundNumber, setRoundNumber] = useState(0)
  const [roundType, setRoundType] = useState<RoundType>(RoundType.Work)
  const [secondsLeft, setSecondsLeft] = useState(1500)
  const [appStateVisible, setAppStateVisible] = useState(true)
  const [showConfirmationModal, setShowConfirmationModal] = useState(false)

  // React Query
  const queryClient = useQueryClient()

  const mutation = useMutation(addRecord, {
    onSuccess: () => {
      void queryClient.invalidateQueries(['records'])
    },
  })

  // ----- TIMER CONTROL METHODS -----
  const advanceRound = useCallback((): void => {
    const nextRound = getNextRound(roundNumber)
    const nextRoundType = getRoundType(nextRound)
    setRoundNumber(nextRound)
    setRoundType(nextRoundType)
    setSecondsLeft(getSecondsReset(nextRoundType))
  }, [roundNumber])

  const startTimer = async (): Promise<void> => {
    const roundData: RoundData = {
      date: new Date().getTime(),
      roundNumber,
      roundType,
    }

    await AsyncStorage.setItem('roundData', JSON.stringify(roundData))
    await scheduleNotification(secondsLeft)
  }

  const startRound = async (): Promise<void> => {
    if (enabled && startButtonEnabled) {
      setTimerActive(true)
      setSecondsLeft(secondsLeft - 1)
      await startTimer()
    }
  }

  const stopRound = async (): Promise<void> => {
    setTimerActive(false)
    setEnabled(false)
    await stopTimer()
  }

  const stopTimer = async (): Promise<void> => {
    const roundData: RoundData | undefined = await getRoundData()

    if (timerActive) {
      advanceRound()
    }

    if (await shouldAddRecord(roundData?.date ?? 0)) {
      await mutation.mutateAsync({
        date: roundData?.date ?? 0,
        type: roundData?.roundType ?? RoundType.Work,
        completed: secondsLeft < 0 ? 1 : 0,
      })
    }

    await AsyncStorage.removeItem('roundData')
    await cancelAllNotifications()
    await setTimeout(() => {
      setSecondsLeft(getSecondsReset(getRoundType(getNextRound(roundNumber - 1))))
      setEnabled(true)
    }, 1000)
  }

  // This effect adds a slight delay to the start button to prevent a known race condition
  useEffect(() => {
    if (enabled) {
      setTimeout(() => {
        setStartButtonEnabled(true)
      }, 1000)
    } else {
      setStartButtonEnabled(false)
    }
  }, [enabled])

  // This effect updates the display time every second
  useEffect(() => {
    const fetchTime = async (): Promise<number> => {
      return await getTimeRemaining()
    }

    if (timerActive && secondsLeft > -1) {
      setTimeout(async () => {
        setSecondsLeft((await fetchTime()) ?? secondsLeft)
      }, 1000)
    } else {
      void (async () => {
        await stopTimer()
      })()
      setTimerActive(false)
      setEnabled(false)
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [secondsLeft])

  const handleAppStateChange = useCallback(
    async (nextAppState: AppStateStatus): Promise<void> => {
      if (nextAppState === 'active') {
        if (!appStateVisible) {
          setTimeout(async () => {
            setSecondsLeft((await getTimeRemaining()) ?? secondsLeft)
            setAppStateVisible(true)
          }, 1000)
        }
      }

      if (nextAppState === 'inactive' || nextAppState === 'background') {
        return setAppStateVisible(false)
      }
    },
    [appStateVisible, secondsLeft],
  )

  // Detecting changes in AppState
  useEffect(() => {
    const subscription = AppState.addEventListener(
      'change',
      async (nextAppState) => await handleAppStateChange(nextAppState),
    )

    return () => {
      subscription.remove()
    }
  }, [handleAppStateChange])

  return (
    <TimerContext.Provider
      value={{
        enabled,
        startButtonEnabled,
        timerActive,
        roundNumber,
        roundType,
        secondsLeft,
        appStateVisible,
        advanceRound,
        startRound,
        stopRound,
        showConfirmationModal,
        setShowConfirmationModal,
      }}
    >
      {children}
    </TimerContext.Provider>
  )
}

export { TimerContext, TimerContextProvider }
