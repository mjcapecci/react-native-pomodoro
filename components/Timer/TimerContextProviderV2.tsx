import React, { createContext, useCallback, useEffect, useState } from 'react'
import { getNextRound, getRoundType, getSecondsReset } from './helpers/timerHelpers'
import { scheduleNotification, cancelAllNotifications } from '../Notifications/notificationManager'
import AsyncStorage from '@react-native-async-storage/async-storage'
import { AppState, AppStateStatus } from 'react-native'
import getTimeRemaining, { getRoundData } from './helpers/timeRemaining'
import { RoundData, RoundType, UserRecord } from '../../types'
import addRecord from '../../data_layer/addRecord'
import shouldAddRecord from './helpers/shouldAddRecord'
import { useMutation, useQueryClient } from '@tanstack/react-query'
import getMostRecentRecord from '../../data_layer/getMostRecentRecord'
import updateRecordCompletedStatus from '../../data_layer/updateRecordCompletedStatus'
import { useToast } from 'react-native-toast-notifications'
import { debounce, throttle } from 'lodash'

export interface TimerContextProps {
  enabled: boolean
  startButtonEnabled: boolean
  timerActive: boolean
  roundNumber: number
  roundType: string
  secondsLeft: number
  appStateVisible: boolean
  advanceRound: () => void
  startRound: () => void
  stopRound: (isPreliminarySkip?: boolean) => void
  showConfirmationModal: boolean
  setShowConfirmationModal: (show: boolean) => void
  showReverseModal: boolean
  setShowReverseModal: (show: boolean) => void
  lastUserRecord: UserRecord | undefined
  reversePreviousRecord: (recordId: string) => void
}

const TimerContext = createContext<TimerContextProps>({
  enabled: true,
  startButtonEnabled: true,
  timerActive: false,
  roundNumber: 0,
  roundType: 'work',
  secondsLeft: 1500,
  appStateVisible: true,
  advanceRound: () => null,
  startRound: () => null,
  stopRound: (isPreliminarySkip) => null,
  showConfirmationModal: false,
  setShowConfirmationModal: () => null,
  showReverseModal: false,
  setShowReverseModal: () => null,
  lastUserRecord: undefined,
  reversePreviousRecord: async (recordId) => null,
})

interface TimerContextProviderProps {
  children: JSX.Element
}

function TimerContextProvider({ children }: TimerContextProviderProps): JSX.Element {
  // ----- STATE -----
  const [enabled, setEnabled] = useState(true)
  const [startButtonEnabled, setStartButtonEnabled] = useState(true)
  const [timerActive, setTimerActive] = useState(false)
  const [roundNumber, setRoundNumber] = useState(0)
  const [roundType, setRoundType] = useState<RoundType>(RoundType.Work)
  const [secondsLeft, setSecondsLeft] = useState(1500)
  const [appStateVisible, setAppStateVisible] = useState(true)
  const [showConfirmationModal, setShowConfirmationModal] = useState(false)
  const [showReverseModal, setShowReverseModal] = useState(false)
  const [lastUserRecord, setLastUserRecord] = useState<UserRecord | undefined>(undefined)

  const toast = useToast()

  // React Query
  const queryClient = useQueryClient()

  const addRecordMutation = useMutation(addRecord, {
    onSuccess: () => {
      void queryClient.invalidateQueries(['records'])
    },
  })

  // ----- TIMER CONTROL METHODS -----
  const advanceRound = useCallback((): void => {
    const nextRound = getNextRound(roundNumber)
    const nextRoundType = getRoundType(nextRound)
    setRoundNumber(nextRound)
    setRoundType(nextRoundType)
    setSecondsLeft(getSecondsReset(nextRoundType))
  }, [roundNumber])

  const startTimer = async (): Promise<void> => {
    const roundData: RoundData = {
      date: new Date().getTime(),
      roundNumber,
      roundType,
    }

    await AsyncStorage.setItem('roundData', JSON.stringify(roundData))
    await scheduleNotification(secondsLeft, roundType)
  }
  const startRound = throttle(async (): Promise<void> => {
    if (enabled && startButtonEnabled) {
      setTimerActive(true)
      setSecondsLeft(secondsLeft - 1)
      await startTimer()
    }
  }, 1000)

  const stopRound = throttle(async (isPreliminarySkip?: boolean): Promise<void> => {
    setTimerActive(false)
    setEnabled(false)
    await stopTimer(isPreliminarySkip ?? false)
  }, 500)

  const stopTimer = throttle(async (isPreliminarySkip: boolean): Promise<void> => {
    const roundData: RoundData | undefined = await getRoundData()

    setShowConfirmationModal(false)

    if (timerActive || isPreliminarySkip) {
      advanceRound()
    }

    if (await shouldAddRecord(roundData?.date ?? 0)) {
      await addRecordMutation.mutateAsync({
        date: roundData?.date ?? 0,
        type: roundData?.roundType ?? RoundType.Work,
        completed: secondsLeft < 0 ? 1 : 0,
      })

      await setLastUserRecord(await getMostRecentRecord())
    }

    if (isPreliminarySkip) {
      await addRecordMutation.mutateAsync({
        date: new Date().getTime(),
        type: roundType,
        completed: 0,
      })
    }

    await AsyncStorage.removeItem('roundData')
    await cancelAllNotifications()

    setSecondsLeft(getSecondsReset(getRoundType(getNextRound(roundNumber - 1))))
    setEnabled(true)
  }, 1000)

  function handleReversePreviousRecord(recordId: string): void {
    void updateRecordCompletedStatus(recordId)
    void queryClient.invalidateQueries(['records'])
    if (lastUserRecord != null) {
      setLastUserRecord({ ...lastUserRecord, completed: 0 })
    }
    toast.show('Previous record reversed.', {
      type: 'success',
      duration: 3000,
      placement: 'top',
      animationType: 'slide-in',
    })
  }

  // This effect adds a slight delay to the start button to prevent a known race condition
  useEffect(() => {
    if (enabled) {
      const enableStartButton = throttle(() => {
        setStartButtonEnabled(true)
      }, 1000)
      enableStartButton()
    } else {
      setStartButtonEnabled(false)
    }
  }, [enabled])

  // This effect updates the display time every second
  useEffect(() => {
    const fetchTime = async (): Promise<number> => {
      return await getTimeRemaining()
    }

    if (timerActive && secondsLeft > -1) {
      const updateSecondsLeft = debounce(async () => {
        setSecondsLeft((await fetchTime()) ?? secondsLeft)
      }, 1000)
      void updateSecondsLeft()
    } else {
      void (async () => {
        await stopTimer(false)
      })()
      setTimerActive(false)
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [secondsLeft])

  const handleAppStateChange = useCallback(
    async (nextAppState: AppStateStatus) => {
      if (nextAppState === 'active') {
        if (!appStateVisible) {
          setSecondsLeft((await getTimeRemaining()) ?? secondsLeft)
          setAppStateVisible(true)
        }
      }
    },
    [appStateVisible, secondsLeft],
  )

  // Detecting changes in AppState
  useEffect(() => {
    const subscription = AppState.addEventListener(
      'change',
      async (nextAppState) => await handleAppStateChange(nextAppState),
    )

    return () => {
      subscription.remove()
    }
  }, [handleAppStateChange])

  return (
    <TimerContext.Provider
      value={{
        enabled,
        startButtonEnabled,
        timerActive,
        roundNumber,
        roundType,
        secondsLeft,
        appStateVisible,
        advanceRound,
        startRound,
        stopRound,
        showConfirmationModal,
        setShowConfirmationModal,
        showReverseModal,
        setShowReverseModal,
        lastUserRecord,
        reversePreviousRecord: handleReversePreviousRecord,
      }}
    >
      {children}
    </TimerContext.Provider>
  )
}

export { TimerContext, TimerContextProvider }
