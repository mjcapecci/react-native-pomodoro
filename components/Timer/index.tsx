import React, { useState, useEffect } from 'react';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { StyleSheet, View } from 'react-native';
import { Heading, Box, Button } from 'native-base';
import Ionicons from '@expo/vector-icons/Ionicons';
import {
  scheduleNotification,
  cancelAllNotifications,
} from '../Notifications/notificationManager';
import {
  fmtMSS,
  getIconColor,
  getNextRound,
  getNextRoundSecondsDisplay,
  getRoundType,
  getSecondsReset,
} from '../../utils/TimerUtils';

const Timer = () => {
  const [enabled, setEnabled] = useState(true);
  const [startButtonEnabled, setStartButtonEnabled] = useState(true);
  const [timerActive, setTimerActive] = useState(false);
  const [roundNumber, setRoundNumber] = useState(-1);
  const [roundType, setRoundType] = useState('work');
  const [secondsLeft, setSecondsLeft] = useState(1500);

  const debugMode = false;

  useEffect(() => {
    if (!timerActive) {
      advanceRound();
    }
  }, [timerActive]);

  useEffect(() => {
    if (enabled) {
      setTimeout(() => {
        setStartButtonEnabled(true);
      }, 1000);
    } else {
      setStartButtonEnabled(false);
    }
  }, [enabled]);

  useEffect(() => {
    if (timerActive && secondsLeft > -1) {
      setTimeout(() => {
        setSecondsLeft(secondsLeft - 1);
      }, 1000);
    } else {
      stopTimer();
      setTimerActive(false);
      setEnabled(false);
    }
  }, [secondsLeft]);

  const advanceRound = () => {
    const nextRound = getNextRound(roundNumber);
    const nextRoundType = getRoundType(nextRound);
    setRoundNumber(nextRound);
    setRoundType(nextRoundType);
    setSecondsLeft(getSecondsReset(nextRoundType));
  };

  const startTimer = async () => {
    const roundData = {
      date: new Date().getTime(),
      roundNumber: roundNumber,
      roundType: roundType,
    };

    await AsyncStorage.setItem('roundData', JSON.stringify(roundData));
    await scheduleNotification(secondsLeft);
  };

  const stopTimer = async () => {
    await AsyncStorage.removeItem('roundData');
    await cancelAllNotifications();
    await setTimeout(() => {
      setSecondsLeft(
        getSecondsReset(getRoundType(getNextRound(roundNumber - 1)))
      );
      setEnabled(true);
    }, 1000);
  };

  const printAsyncStorage = async () => {
    const startTime = await AsyncStorage.getItem('roundData');
    console.log(startTime);
  };

  return enabled ? (
    <View style={styles.container}>
      <Box style={styles.timeHeaderContainer}>
        <Heading style={styles.timeHeader}>
          {timerActive
            ? secondsLeft > 0
              ? fmtMSS(secondsLeft)
              : '0:00'
            : getNextRoundSecondsDisplay(roundNumber)}
        </Heading>
      </Box>
      <Box style={styles.starContainer}>
        <Ionicons
          name='hammer'
          color={getIconColor(0, roundNumber, timerActive)}
          size={24}
          style={styles.star}
        />
        <Ionicons
          name='walk'
          color={getIconColor(1, roundNumber, timerActive)}
          size={24}
          style={styles.star}
        />
        <Ionicons
          name='hammer'
          color={getIconColor(2, roundNumber, timerActive)}
          size={24}
          style={styles.star}
        />
        <Ionicons
          name='walk'
          color={getIconColor(3, roundNumber, timerActive)}
          size={24}
          style={styles.star}
        />
        <Ionicons
          name='hammer'
          color={getIconColor(4, roundNumber, timerActive)}
          size={24}
          style={styles.star}
        />
        <Ionicons
          name='pizza'
          color={getIconColor(5, roundNumber, timerActive)}
          size={24}
          style={styles.star}
        />
      </Box>
      <Box style={styles.actionButton}>
        {!timerActive ? (
          <Ionicons
            name='play-circle-outline'
            color={startButtonEnabled ? 'white' : 'grey'}
            size={104}
            onPress={() => {
              if (enabled && startButtonEnabled) {
                setTimerActive(true);
                setSecondsLeft(secondsLeft - 1);
                startTimer();
              }
            }}
          ></Ionicons>
        ) : (
          <Ionicons
            name='stop-circle-outline'
            color='white'
            size={104}
            onPress={() => {
              setTimerActive(false);
              setEnabled(false);
              stopTimer();
            }}
          ></Ionicons>
        )}
      </Box>
      <Button style={!timerActive ? styles.skipButton : styles.hideSkip}>
        {' '}
        <Ionicons
          name='play-skip-forward-circle'
          color='white'
          size={50}
          onPress={() => advanceRound()}
        ></Ionicons>
      </Button>
      {debugMode && (
        <Button>
          {' '}
          <Ionicons
            name='play-skip-forward-circle'
            color='white'
            size={50}
            onPress={() => printAsyncStorage()}
          ></Ionicons>
        </Button>
      )}
    </View>
  ) : (
    <View style={styles.container}>
      <Box style={styles.timeHeaderContainer}>
        <Heading>Loading...</Heading>
      </Box>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: '#000',
  },
  timeHeaderContainer: {
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
  },
  timeHeader: {
    fontSize: 84,
    color: '#fff',
    marginBottom: 5,
    fontVariant: ['tabular-nums'],
  },
  starContainer: {
    display: 'flex',
    flexDirection: 'row',
  },
  star: {
    margin: 3,
  },
  actionButton: {
    marginTop: 60,
  },
  skipButton: {
    marginTop: 30,
    backgroundColor: '#000',
  },
  hideSkip: {
    display: 'none',
  },
});

export default Timer;
