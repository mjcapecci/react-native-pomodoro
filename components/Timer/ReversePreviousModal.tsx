import * as React from 'react'
import { Modal, Portal, Text, Button } from 'react-native-paper'
import { TimerContext } from './TimerContextProvider'

interface ReversePreviousModalProps {
  showModal: boolean
  setShowModal: (showModal: boolean) => void
}

const ReversePreviousModal = ({
  showModal,
  setShowModal,
}: ReversePreviousModalProps): JSX.Element => {
  const time = React.useContext(TimerContext)

  const hideModal = (): void => setShowModal(false)
  const containerStyle = { backgroundColor: 'white', padding: 20 }

  const confirmReverse = async (): Promise<void> => {
    if (time?.lastUserRecord?.id != null) {
      await time.reversePreviousRecord(time.lastUserRecord.id)
    }
    hideModal()
  }

  return (
    <Portal>
      <Modal visible={showModal} onDismiss={hideModal} contentContainerStyle={containerStyle}>
        <Text>
          Are you sure you would like to set the last completed round to a status of skipped?
        </Text>
        <Button onPress={async () => await confirmReverse()}>Skip</Button>
        <Button onPress={() => hideModal()}>Cancel</Button>
      </Modal>
    </Portal>
  )
}

export default ReversePreviousModal
