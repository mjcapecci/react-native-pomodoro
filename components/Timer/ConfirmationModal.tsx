import * as React from 'react'
import { Modal, Portal, Text, Button } from 'react-native-paper'
import { TimerContext } from './TimerContextProviderV2'
import styles from '../General/General.style'
import ModalDivider from '../General/ModalDivider'

interface ConfirmationModalProps {
  skipType: 'skip' | 'stop'
  showModal: boolean
  setShowModal: (showModal: boolean) => void
  secondsLeft: number
}

const ConfirmationModal = ({
  skipType,
  showModal,
  setShowModal,
  secondsLeft,
}: ConfirmationModalProps): JSX.Element => {
  const time = React.useContext(TimerContext)

  const hideModal = (): void => setShowModal(false)
  const containerStyle = { backgroundColor: 'white', padding: 20 }

  const confirmSkip = (): void => {
    if (skipType === 'stop') {
      time.stopRound()
    } else {
      time.stopRound(true)
    }
    hideModal()
  }

  const disabled = secondsLeft < 3

  return (
    <Portal>
      <Modal visible={showModal} onDismiss={hideModal} contentContainerStyle={containerStyle}>
        <Text style={styles.centeredModalText}>
          Are you sure you would like to skip this round?
        </Text>
        <ModalDivider />
        <Button onPress={() => hideModal()}>Cancel</Button>

        <ModalDivider />
        <Button
          style={styles.bottomButton}
          color={disabled ? 'grey' : '#ed0d0dc4'}
          onPress={() => !disabled && confirmSkip()}
        >
          Skip {disabled && '(not available)'}
        </Button>
      </Modal>
    </Portal>
  )
}

export default ConfirmationModal
