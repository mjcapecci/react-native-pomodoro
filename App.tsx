import React, { useEffect, useMemo } from 'react'
import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import { createBottomTabNavigator } from '@react-navigation/bottom-tabs'
import { DarkTheme, NavigationContainer } from '@react-navigation/native'
import { Provider as PaperProvider, DarkTheme as DefaultTheme } from 'react-native-paper'
import * as ScreenOrientation from 'expo-screen-orientation'
import { ToastProvider } from 'react-native-toast-notifications'
import Constants from 'expo-constants'
import Ionicons from '@expo/vector-icons/Ionicons'

import Timer from './components/Timer'
import Stats from './components/Stats'
import DevTools from './components/DevTools'

import createTable from './data_layer/createTable'

import * as Notifications from 'expo-notifications'
import {
  askPermissions,
  cancelAllNotifications,
  getNotifications,
} from './components/Notifications/notificationManager'
import { TimerContextProvider } from './components/Timer/TimerContextProvider'
import { AppContextProvider } from './components/General/AppContextProvider'
import VersionModal from './components/General/VersionModal'
import Info from './components/Info'

void ScreenOrientation.lockAsync(ScreenOrientation.OrientationLock.PORTRAIT_UP)

Notifications.setNotificationHandler({
  handleNotification: async () => ({
    shouldShowAlert: true,
    shouldPlaySound: true,
    shouldSetBadge: false,
  }),
})

// Create a client
const queryClient = new QueryClient()

const App = (): JSX.Element => {
  const Tab = createBottomTabNavigator()
  const enableDevTools = Constants?.manifest?.extra?.enableDevTools

  useEffect(() => {
    async function initializeSQLite(): Promise<void> {
      return await createTable()
    }

    initializeSQLite().catch((err) => console.log(err))
  }, [])

  // get initial notification permissions if not exist
  useEffect(() => {
    async function getInitialPermissions(): Promise<Notifications.NotificationPermissionsStatus> {
      return await askPermissions()
    }

    getInitialPermissions().catch((err) => console.log(err))
  })

  // deletes existing notifications if any exist after full recycle of app
  useMemo(() => {
    void (async () => {
      const notifications = await getNotifications()
      if (notifications.length !== 0) {
        console.log('Notifications Canceled')
        void cancelAllNotifications()
      }
    })()
  }, [])

  const theme = {
    ...DefaultTheme,
    colors: {
      ...DefaultTheme.colors,
      primary: '#616161',
    },
  }

  return (
    <QueryClientProvider client={queryClient}>
      <AppContextProvider>
        <PaperProvider theme={theme}>
          <ToastProvider offset={50}>
            <TimerContextProvider>
              <>
                <NavigationContainer theme={DarkTheme}>
                  <Tab.Navigator screenOptions={{ headerShown: true }}>
                    <Tab.Screen
                      name='Timer'
                      component={Timer}
                      options={{
                        tabBarShowLabel: false,
                        tabBarActiveTintColor: 'gold',
                        tabBarIcon: ({ focused, color }) => (
                          <Ionicons
                            focused={focused}
                            name='timer-outline'
                            size={22}
                            color={focused ? color : 'white'}
                          />
                        ),
                      }}
                    />
                    <Tab.Screen
                      name='Stats'
                      component={Stats}
                      options={{
                        tabBarShowLabel: false,
                        tabBarActiveTintColor: 'gold',
                        tabBarIcon: ({ focused, color }) => (
                          <Ionicons
                            focused={focused}
                            name='stats-chart'
                            size={22}
                            color={focused ? color : 'white'}
                          />
                        ),
                      }}
                    />
                    <Tab.Screen
                      name='Help'
                      component={Info}
                      options={{
                        tabBarShowLabel: false,
                        tabBarActiveTintColor: 'gold',
                        title: 'Info',
                        tabBarIcon: ({ focused, color }) => (
                          <Ionicons
                            focused={focused}
                            name='information-circle'
                            size={22}
                            color={focused ? color : 'white'}
                          />
                        ),
                      }}
                    />
                    {Boolean(enableDevTools) && (
                      <Tab.Screen
                        name='Tools'
                        component={DevTools}
                        options={{
                          tabBarShowLabel: false,
                          tabBarActiveTintColor: 'gold',
                          tabBarIcon: ({ focused, color }) => (
                            <Ionicons
                              focused={focused}
                              name='code-outline'
                              size={22}
                              color={focused ? color : 'white'}
                            />
                          ),
                        }}
                      />
                    )}
                  </Tab.Navigator>
                </NavigationContainer>
                <VersionModal />
              </>
            </TimerContextProvider>
          </ToastProvider>
        </PaperProvider>
      </AppContextProvider>
    </QueryClientProvider>
  )
}

export default App
