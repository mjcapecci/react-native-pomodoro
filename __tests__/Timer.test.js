import React from 'react';
import renderer from 'react-test-renderer';

import App from '../App.tsx';
import Timer from '../components/Timer/index';

describe('<Timer />', () => {
  it('to match screenshot', () => {
    const tree = renderer
      .create(
        <App>
          <Timer />
        </App>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
