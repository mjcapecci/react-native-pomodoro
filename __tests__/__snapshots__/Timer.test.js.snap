// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Timer /> to match screenshot 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
/>
`;
